### PR TITLE
More code review feedback

### PR DESCRIPTION
**Pull request by ctubbsii**
This feedback includes:

* Ensure that a client-side warning is generated, as well as a
  server-side warning for the system logs, when setting/removing
  individual properties in the system configuration in ZooKeeper
* Limit the support for handling property renames to just those in the
  system configuration, and not in the table or namespace configuration,
  because those are not accompanied by corresponding upgrade code, and
  updating those could have unintended consequences in future if not
  handled in a way that is possibly specific to future changes. Since we
  can't fully anticipate those future possibilities, it's easier to
  narrow the scope of this work to just handling the system properties
  that exist now.
* Remove tracking of previously logged warnings, and instead warn on
  every new attempt to set or use a system property that is deprecated
  and has a replacement. That way, the log message isn't lost to rolled
  over logs, or hidden from long running clients (like the shell, for
  example), etc.
* Declare renamers to be an ordered list and rename iteratively,
  applying each renamer in turn. This isn't used today, but might be
  useful in future.
* Allow for custom logging messages and/or other custom actions when a
  replacement name is identified for a deprecated property. This allows
  for code reuse of replacement name lookups, even if the log message
  needs to be different.
* Change the way the sanity check works for the SiteConfiguration to
  ensure that users aren't mixing and matching the old master properties
  with the new manager ones in the same configuration, since that could
  get complicated to reason about which take precedence. If a user
  updates their config files to rename `master.*` to `manager.*`, they
  must now do so for all matching properties, or none.
* Simplify isValidZooPropertyKey method by matching on the old property
  name, without needing to actually perform the replacement mapping.
* Fail safe by *not* automatically remapping a deprecated property that
  is requested to be removed. It is hard to determine the user's intent
  here, as they may simply be trying to remove their own use of
  deprecated properties, and may not actually want to remove the
  property with the new name that replaced it. So, warn, but don't take
  action, when requesting to remove a deprecated property from ZooKeeper
  that shouldn't exist there after an upgrade, thereby only permitting
  removal of a property by the name that actually exists in ZooKeeper.
* Simplify Upgrader code using getAllPropertiesWithPrefix, and with a
  custom remove method in SystemPropUtil that won't generate a warning
  when attempting remove a property using the deprecated name (custom
  code just for upgrade).
* Drop test cases that don't make sense with these proposed behavior
  changes, and improve remaining tests. Rename Upgrade IT to follow
  other Upgrade IT naming convention to show which data version it
  upgrades to/from.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java [hunk 1]
```diff
@@ -76,12 +83,20 @@ public void setProperty(final String property, final String value)
   public void removeProperty(final String property)
       throws AccumuloException, AccumuloSecurityException {
     checkArgument(property != null, "property is null");
+    DeprecatedPropertyUtil.getReplacementName(property, (log, replacement) -> {
+      // force a warning on the client side, but send the name the user used to the server-side
+      // to trigger a warning in the server logs, and to handle it there
+      log.warn("{} was deprecated and will be removed in a future release;"
+          + " no action was taken because it is not set here;"
+          + " did you mean to remove its replacment {} instead?", property, replacement);
```
Comment (brianloss) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/InstanceOperationsImpl.java:91]: See related comment on SystemPropUtil.removeSystemProperty.

File: core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java [hunk 2]
```diff
@@ -21,72 +21,76 @@
...
+  public static String getReplacementName(final String propertyName,
+      BiConsumer<Logger,String> loggerActionOnReplace) {
+    String replacement = propertyName;
+    for (PropertyRenamer renamer : renamers) {
+      if (renamer.keyFilter.test(replacement)) {
+        replacement = renamer.keyMapper.apply(replacement);
+      }
+    }
+    // perform the logger action if the property was replaced
+    if (loggerActionOnReplace != null && !replacement.equals(propertyName)) {
```
Comment (brianloss) [inline core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java:90]: It looks like the only place a null is passed is in the unit test. I'm wondering if it would be better to enforce a non-null logger action here, and add a protected method for tests that doesn't take a logger action? That method would delegate to this one passing an empty lambda. Either that, or maybe change the tests to pass a do-nothing lambda (via a constant) to make it more readable--to get rid of all the null params and instead have a name with meaning.

**Review by brianloss — CHANGES_REQUESTED** [review]
Look great, thanks! I had a few questions and minor issues before merging...

File: core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java [hunk 3]
```diff
@@ -96,15 +100,19 @@ public static String renameDeprecatedProperty(String propertyName) {
...
+      if (MASTER_MANAGER_RENAMER.keyFilter.test(p)) {
+        foundMasterPrefix.set(true);
+      } else if (p.startsWith(Property.MANAGER_PREFIX.getKey())) {
+        foundManagerPrefix.set(true);
       }
     });
+    if (foundMasterPrefix.get() && foundManagerPrefix.get()) {
+      throw new IllegalStateException(
+          "Found both old 'master.*' and new 'manager.*' naming conventions in the same startup configuration");
+    }
```
[the remark this answers is not in the document]
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java:116]: I like that better.

File: core/src/main/java/org/apache/accumulo/core/conf/SiteConfiguration.java [hunk 4]
```diff
@@ -168,11 +168,16 @@ public SiteConfiguration build() {
...
 
       var result = new HashMap<String,String>();
-      config.getKeys().forEachRemaining(k -> result
-          .put(DeprecatedPropertyUtil.renameDeprecatedPropertyAndWarn(k), config.getString(k)));
+      config.getKeys().forEachRemaining(original -> {
+        String actual = DeprecatedPropertyUtil.getReplacementName(original, (log, replacement) -> {
+          log.warn("{} has been deprecated and will be removed in a future release;"
+              + " loading as its replacement {} instead.", original, replacement);
+        });
+        result.put(actual, config.getString(original));
```
[the remark this answers is not in the document]
Comment (brianloss) [reply core/src/main/java/org/apache/accumulo/core/conf/SiteConfiguration.java:179]: I misread this code. Scratch this comment...

File: core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java [hunk 5]
```diff
@@ -21,72 +21,76 @@
 import static java.util.Objects.requireNonNull;
 
 import java.util.ArrayList;
-import java.util.HashSet;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
 import java.util.function.Predicate;
 import java.util.function.UnaryOperator;
```
Comment (brianloss) [inline core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java:28]: Imports needed for my previous (accepted) suggestion...
```suggestion
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.BiConsumer;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;
import java.util.stream.StreamSupport;
```

File: core/src/main/java/org/apache/accumulo/core/conf/SiteConfiguration.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/conf/SiteConfiguration.java:179]: `result` started as an empty map... it only contains actual properties resolved from this replacement. Only properties that are not subject to a rename, and properties that are subject to a rename *after* they've been renamed, ever get inserted into this map. This is the same behavior as before. I just split it into two statements, an assignment and a `put()` on separate lines in the `forEachRemaining` lambda, rather than have it as one line, which was slightly harder for me to follow after adding the lambda for the custom warning.

File: server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java [hunk 6]
```diff
@@ -75,7 +81,20 @@ public static void setSystemProperty(ServerContext context, String property, Str
...
+    DeprecatedPropertyUtil.getReplacementName(property, (log, replacement) -> {
+      log.warn("{} was deprecated and will be removed in a future release;"
+          + " no action was taken because it is not set here;"
+          + " did you mean to remove its replacment {} instead?", property, replacement);
+      shouldRemove.set(false);
+    });
+    if (shouldRemove.get()) {
+      removePropWithoutDeprecationWarning(context, property);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java:94]: Continuing to allow them to set the property makes sense (we will just remap it to the new name). Continuing to allow them to remove the property also makes sense, but there's nothing to do, since that property won't exist in ZooKeeper. Automatically remapping to the new property on remove seems dangerous to me, because we can't know the user's intent. If they say "remove X" , and we automatically instead "remove Y", that doesn't seem to be good behavior to me. I'd rather emit the warning, asking them if they are sure. It's safer than guessing user intent.

File: core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java (hunk 2 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/conf/DeprecatedPropertyUtil.java:92]: I can pass a NOOP action for the tests.

File: server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (brianloss) [reply server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java:94]: Fair point. I'm not sure how strongly I feel about this issue, but I could say that doing nothing on remove is also making a guess on the user's intent. If I take the view of a user who says "I'm not ready to update any property names yet, and I'm going to keep using the old properties." then doing nothing on remove breaks remove behavior. The old name isn't really just deprecated since I cannot use it on remove calls.

I suppose the argument the other way (if remove does take action with the renamed property) is if I'm a user trying to take care of renaming and I think I need to remove the old ZK property as a part of that effort, then we'd actually be removing the new one. In that case, the user taking the action would see the warning message and would know they need to add the new property back (though maybe we need to log the property value as well as the name if that's the case).

[the remark this answers is not in the document]
Comment (brianloss) [reply server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java:94]: I know one could also argue the same thing about seeing a log message if the old name is used and no action is taken. I was thinking about the automation case that Keith mentioned previously. In that case, if something automated is using the old name, then I would say we're not guessing intent--it was to use the new name but automation hasn't been updated yet.

[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java:94]: Of the choices, I think not removing something that doesn't exist is safest. I suppose in some sense, "doing nothing" could be interpreted as inferring user intent, but I think that's a stretch. We are literally taking no action one way or the other. Even if that's not an entirely neutral choice, it's as neutral as we can possibly get, and it avoids accidental deletions of config, so it's safer. But, given that it could be the user's intent to delete the new property, I went with being as loud and verbose with the warning as possible, so the user knew *explicitly* what action was taken or not taken.

We can always call this behavior out in an upgrade section of the release notes.

[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java:94]: @keith-turner had an opinion about the automation. Perhaps he has an opinion about this and wants to weigh in before it's merged into your branch? (Alternatively, could merge it in to the branch, and request a re-review there, since it might be nice to take a look at the overall change in context.)

[the remark this answers is not in the document]
Comment (brianloss) [reply server/base/src/main/java/org/apache/accumulo/server/util/SystemPropUtil.java:94]: I think it's better to merge this in and then ask @keith-turner to re-review. There are enough changes now that he might have opinions on other sections.

**Review by brianloss — APPROVED** [review]
